# Build notes part 2

## Handling stray api requests

If you have a request that makes it past all your routes you should then have a catch-all route to respond with a 404

Instead of declaring a bunch of `app.verbs` you can just use `app.all()`

```javascript
app.all('*', (request, response, next) => {
    response.status(404).json({
        status: 'fail',
        message: `Route ${request.originalUrl} does not exist`,
    });
});
```

## Express error handling

Important to distinguish btwn two types of errors

-   Operational
-   Programming

Operational errors are ones that we can predict and control for, usually this boils to everything that is not PEBKAC

Programming errors are basically PEBKAC type stuff

Also, exceptions and errors are interchangeable

Express actually comes with error handling out of the box, so we can just write a global middleware to handle our error signaling that can accept error signals from anywhere in the app

Global error handling middleware enables us to maintain good separation of concerns btwn business logic and applicatin logic

Here is an example of a better error handling mechanism than the one above (in app.js)

```javascript
// Controlling for a route that isnt handled by any router listed above
app.all('*', (request, response, next) => {
    // response.status(404).json({
    //   status: 'fail',
    //   message: `Route ${request.originalUrl} does not exist`,
    // });

    const error = new Error(
        `Requested route ${request.originalUrl} does not exist`
    );
    error.status = 'fail';
    error.statusCode = 404;

    // passing an error to next() will cause it to go to our error handler
    next(error);
});

app.use((error, request, response, next) => {
    // eslint-disable-next-line no-param-reassign
    error.statusCode = error.statusCode || 500;
    // eslint-disable-next-line no-param-reassign
    error.status = error.status || 'Unknown error';

    response.status(error.statusCode).json({
        status: error.status,
        message: error.message,
    });
});
```

# Making a custom Error class

```javascript
class AppError extends Error {
    constructor(message, statusCode) {
        // calling the parent class which already has a message
        super(message);

        this.statusCode = statusCode;
        // of the status code starts with a 4 its a fail otherwise its error
        this.status = `${statusCode}`.startsWith('4') ? 'fail' : 'error';
        // setting this to be just for operational errors
        this.isOperational = true;

        // adding a stack trace
        Error.captureStackTrace(this, this.constructor);
    }
}

module.exports = AppError;
```

```javascript
// Controlling for a route that isnt handled by any router listed above
app.all('*', (request, response, next) => {
    next(
        new AppError(
            `Requested route ${request.originalUrl} does not exist`,
            404
        )
    );
});
```

making a custom controller for errors

```javascript
/* eslint-disable no-param-reassign */
module.exports = (error, request, response, next) => {
    error.statusCode = error.statusCode || 500;
    error.status = error.status || 'Unknown error';

    response.status(error.statusCode).json({
        status: error.status,
        message: error.message,
    });
};
```

import the new controller up top then use it like so

```javascript
// Controlling for a route that isnt handled by any router listed above
app.all('*', (request, response, next) => {
    // const error = new Error(
    //   `Requested route ${request.originalUrl} does not exist`
    // );
    // error.status = 'fail';
    // error.statusCode = 404;

    // passing an error to next() will cause it to go to our error handler
    next(
        new AppError(
            `Requested route ${request.originalUrl} does not exist`,
            404
        )
    );
});

app.use(globalErrorHandler);
```

## Wordy explanation of the refactor

At the bottom of our app.js we create an `app.all()` to handle any route requests that dont get handled by any of our actual routers (effectively, a default).

```javascript
app.all('*', (request, response, next) => {
    next(
        new AppError(
            `Requested route ${request.originalUrl} does not exist`,
            404
        )
    );
});
```

In that, we assume that since no router higher up was hit, the request is invalid. In the callback function we simply call next() and generate a new AppError. AppError is a custom class that we have created to handle errors easier for us as well as to enable us to capture stack traces on our exceptions.

```javascript
class AppError extends Error {
    constructor(message, statusCode) {
        // calling the parent class which already has a message
        super(message);

        this.statusCode = statusCode;
        // of the status code starts with a 4 its a fail otherwise its error
        this.status = `${statusCode}`.startsWith('4') ? 'fail' : 'error';
        // setting this to be just for operational errors
        this.isOperational = true;

        // adding a stack trace
        Error.captureStackTrace(this, this.constructor);
    }
}

module.exports = AppError;
```

With this class we can create a custom controller to quickly spawn errors

```javascript
/* eslint-disable no-param-reassign */
module.exports = (error, request, response, next) => {
    error.statusCode = error.statusCode || 500;
    error.status = error.status || 'Unknown error';

    response.status(error.statusCode).json({
        status: error.status,
        message: error.message,
    });
};
```

We then use this controller as the route handler at the bottom of our `app.js`

```javascript
app.use(errorController);
```

## Removing try catch blocks from controllers

Simply wrap async function inside another function, which must return an anonymous function so that it can be reassigned to your original route export, otherwise the catch function executes on compile and cant access `request`, `response` or `next`

Before:

```javascript
exports.createTour = async (request, response) => {
    try {
        // Creating an instance of a Tour and then calling its methods
        // const newTour = new Tour({});
        // newTour.save();

        // Rawdogging it and calling the method off the parent
        const newTour = await Tour.create(request.body);

        response.status(201).json({
            status: 'success',
            data: {
                tour: newTour,
            },
        });
    } catch (error) {
        response.status(400).json({
            status: 'fail',
            message: 'Invalid data sent',
            details: error,
        });
    }
};
```

After:

```javascript
const catchAsync = fn => (request, response, next) => {
    fn(request, response, next).catch(error => next(error));
};

exports.createTour = catchAsync(async (request, response, next) => {
    const newTour = await Tour.create(request.body);

    response.status(201).json({
        status: 'success',
        data: {
            tour: newTour,
        },
    });
});
```

You can also use the catchAsync helper in the router file too, but it makes it harder to track whats async and what isnt

Example of throwing an error in a handler function with this new method

```javascript
exports.getTour = catchAsync(async (request, response, next) => {
    const tour = await Tour.findById(request.params.id);

    if (!tour) {
        return next(new AppError('Requested tour does not exist', 404));
    }

    response.json({
        status: 'success',
        params: request.params,
        data: {
            tour,
        },
    });
});
```

Separating operational/trusted errors from errors not caused by us

```javascript
const sendErrorForUser = (error, response) => {
    if (error.isOperational) {
        response.status(error.statusCode).json({
            status: error.status,
            message: error.message,
        });
    } else {
        response.status(500).json({
            status: 'error',
            message: `Unknown error`,
        });
    }
};
```

Important to note that you also need to make errors from trusted platforms operational everywhere, so that mongoose errors dont throw a generic error for instance

```javascript
const handleCastErrorDB = error => {
    const message = `Invalid: ${error.path} is ${error.value}`;
    return new AppError(message, 400);
};

const handleDuplicateFieldsDB = error => {
    const duplicateFields = Object.keys(error.keyPattern).join(', ');
    const message = `Duplicate field values for the following: ${duplicateFields}. please use another value`;
    return new AppError(message, 400);
};

const handleValidationErrorDB = error => {
    const invalidFields = Object.values(error.errors).map(
        element => element.message
    );
    const message = `Invalid input data. ${invalidFields.join('. ')}.`;
    return new AppError(message, 400);
};
```

Entire error controller

```javascript
const AppError = require('../utils/appError');

/* eslint-disable no-param-reassign */
// Error generators
const handleCastErrorDB = error => {
    const message = `Invalid: ${error.path} is ${error.value}`;
    return new AppError(message, 400);
};

const handleDuplicateFieldsDB = error => {
    const duplicateFields = Object.keys(error.keyPattern).join(', ');
    const message = `Duplicate field values for the following: ${duplicateFields}. please use another value`;
    return new AppError(message, 400);
};

const handleValidationErrorDB = error => {
    const invalidFields = Object.values(error.errors).map(
        element => element.message
    );
    const message = `Invalid input data. ${invalidFields.join('. ')}.`;
    return new AppError(message, 400);
};

// Error handlers
const sendErrorForDev = (error, response) => {
    response.status(error.statusCode).json({
        status: error.status,
        message: error.message,
        stack: error.stack,
        error: { ...error, name: error.name },
    });
};

const sendErrorForProd = (error, response) => {
    if (error.isOperational) {
        response.status(error.statusCode).json({
            status: error.status,
            message: error.message,
        });
    } else {
        // eslint-disable-next-line no-console
        console.log(`🤖 ERROR 🤖: ${error}`);
        response.status(500).json({
            status: 'error',
            message: `Unknown error, probably caused by a dependancy.`,
        });
    }
};

// Core function
module.exports = (error, request, response, next) => {
    error.statusCode = error.statusCode || 500;
    error.status =
        error.status ||
        'Unhandled error 😬, you should not be able to see this if you are in prod';

    if (process.env.NODE_ENV === 'development') {
        // sending error in development
        sendErrorForDev(error, response);
    } else if (process.env.NODE_ENV === 'production') {
        // sending errors in production

        // for some goddamn reason you have to spread error, and then create a name
        // property with a value of error.name, otherwise the property isnt readable
        let err = { ...error, name: error.name };
        if (err.name === 'CastError') err = handleCastErrorDB(err);
        if (err.code === 11000) err = handleDuplicateFieldsDB(err);
        if (err.name === 'ValidationError') err = handleValidationErrorDB(err);

        sendErrorForProd(err, response);
    }
};
```

# Handling unhandled promise rejections

```javascript
// handling promise rejections
process.on('unhandledRejection', error => {
    console.log(error.name, error.message);
});
```

or a bit more complex

```javascript
const server = app.listen(port, () => {
    // eslint-disable-next-line no-console
    console.log(`App running on port ${port}...`);
});

// handling promise rejections
process.on('unhandledRejection', error => {
    console.warn(error.name, error.message);
    console.error(`UNHANDLED REJECTION, SHUTTING DOWN...`);
    server.close(() => {
        process.exit(1);
    });
});
```

errors spawned in middleware will get passed off to express error handling middleware

# Authentication, Authorization and Security

We will be using JSON web tokens for authentication

Basic signup function

```javascript
const User = require('../models/userModel');
const catchAsync = require('../utils/catchAsync');

exports.signUp = catchAsync(async (request, response, next) => {
    const newUser = await User.create(request.body);

    response.status(201).json({
        status: 'success',
        data: { user: newUser },
    });
});
```

This is a route that is user acessible, but its also important to have/use admin
tools where necessary

Passwords should never ever be stored in cleartext!!

User model

```javascript
const mongoose = require('mongoose');
const validator = require('validator');

const userSchema = new mongoose.Schema({
    // name, email, photo, password, passwordconfirm
    name: {
        type: String,
        required: [true, 'No name specified'],
    },
    email: {
        type: String,
        required: [true, 'No E-Mail specified'],
        unique: true,
        lowercase: true,
        validate: [validator.isEmail, 'E-Mail is invalid'],
    },
    photo: {
        type: String,
    },
    password: {
        type: String,
        required: [true, 'No password specified'],
        minlength: [8, 'Password is too short'],
    },
    passwordConfirm: {
        type: String,
        required: [true, 'Must confirm password'],
        // this validator only fires on save!
        validate: {
            validator(el) {
                return el === this.password;
            },
        },
        message: 'Passwords must match',
    },
});

const User = mongoose.model('User', userSchema);

module.exports = User;
```

## Encrypting passwords

```javascript
userSchema.pre('save', async function (next) {
    // only run if pw is modified
    if (!this.isModified('password')) return next();

    // hash pw with cost of 12
    this.password = await bcrypt.hash(this.password, 12);

    // delete passwordConfirm as it is unneeded now
    this.passwordConfirm = undefined;
});
```

# Authenticating with JWT

There are many auth methods but this one is modern, simple and secure

Stateless solution for auth

No need to store anything in memory on server

Most common alternative is to maintain sessions on the server for logged in users

Login:

-   Client makes post request with email and pw
-   If user and password exist and are correct, a unique JWT is created
-   JWT is secret
-   Server responds with JWT which is stored in cookie or local storage
-   A JWT is effectively a ticket-granting ticket a la Kerberos
-   JWT is sent along with the request
-   If server determines the JWT is valid, access is allowed
-   Server responds with protected data
-   All of this communication must happen over HTTPS, otherwise the JWT can get MITM'd

JWT's use HMAC-SHA256

Header and Payload are encoded not encrypted  
Signature is created with header payload and secret

Signing alg takes the header payload and secret to create the signature, which all 3 form the JWT

Signature acts as an integrity check for the rest of the JWT

-   Upon receiving JWT, server will take header and payload from incoming JWT
-   Together with its known secret, it will create a test signature
-   The test signature is compared to the original signature
-   If they match the user is authenticated

Secret protection is paramount because it's what keeps connections and users secure

Auth is very hard to get right :(

It's a complex topic and easy to fuck up

We're gonna raw dog everything except JWT compute

There is a problem with the old signUp function we wrote

```javascript
const User = require('../models/userModel');
const catchAsync = require('../utils/catchAsync');

exports.signUp = catchAsync(async (request, response, next) => {
    const newUser = await User.create(request.body);

    response.status(201).json({
        status: 'success',
        data: { user: newUser },
    });
});
```

In this implementation, anyone can give themselves admin

Fixed version below

```javascript
const User = require('../models/userModel');
const catchAsync = require('../utils/catchAsync');

exports.signUp = catchAsync(async (request, response, next) => {
    const userToCreate = {
        name: request.body.name,
        email: request.body.email,
        password: request.body.password,
        passwordConfirm: request.body.passwordConfirm,
    };

    const newUser = await User.create(userToCreate);

    response.status(201).json({
        status: 'success',
        data: { user: newUser },
    });
});
```

This strips us of the ability to register anyone as an admin

You now have to grant people admin through compass

Creating users with JWT's  
This is handy for when you want to log someone in right away

```javascript
const jwt = require('jsonwebtoken');
const User = require('../models/userModel');
const catchAsync = require('../utils/catchAsync');

exports.signUp = catchAsync(async (request, response, next) => {
    const userToCreate = {
        name: request.body.name,
        email: request.body.email,
        password: request.body.password,
        passwordConfirm: request.body.passwordConfirm,
    };

    const newUser = await User.create(userToCreate);

    const token = jwt.sign({ id: newUser._id }, process.env.JWT_SECRET, {
        expiresIn: process.env.JWT_EXPIRES_IN,
    });

    response.status(201).json({
        status: 'success',
        token,
        data: { user: newUser },
    });
});
```

# Logging users in

In this case we log users in by evaluating credentials and responding with JWT if creds are correct

Side note: set `select: false` on the password field in the model in order to not return it from the api

This also spills over to the auth controller and not accidentally leaking the hashed password to the client

In order to get a prop you know is there, write `.select('+thingyourelookingfor')

To compare passwords make use of bcrypt again and an instance method

Instance methods are methods that are available on all documents of a collection

```javascript
// in the model
userSchema.methods.correctPassword = async function (
    candidatePassword,
    userPassword
) {
    return await bcrypt.compare(candidatePassword, userPassword);
};
```

completed auth controller

```javascript
const jwt = require('jsonwebtoken');
const User = require('../models/userModel');
const catchAsync = require('../utils/catchAsync');
const AppError = require('../utils/appError');

// helper function to sign a token
const signToken = id =>
    jwt.sign({ id }, process.env.JWT_SECRET, {
        expiresIn: process.env.JWT_EXPIRES_IN,
    });

exports.signUp = catchAsync(async (request, response, next) => {
    const userToCreate = {
        name: request.body.name,
        email: request.body.email,
        password: request.body.password,
        passwordConfirm: request.body.passwordConfirm,
    };

    const newUser = await User.create(userToCreate);

    const token = signToken(newUser._id);

    response.status(201).json({
        status: 'success',
        token,
        data: { user: newUser },
    });
});

exports.login = catchAsync(async (request, response, next) => {
    const { email, password } = request.body;
    // 1) Check if email and passwords exist in the request
    if (!email || !password) {
        return next(new AppError('No e-mail or password specified', 400));
    }

    // 2) Check if the user exists
    const user = await User.findOne({ email }).select('+password');

    // if there is no user or the evaluation of passwords fails, send an error
    if (!user || !(await user.correctPassword(password, user.password))) {
        return next(new AppError('The username or password is invalid', 401));
    }

    // 3) If everything ok respond with JWT with users id in this case
    const token = signToken(user._id);
    response.json({
        status: 'success',
        token,
    });
});
```

Part 2 of implementing logins is to actually protect routes

This can be achieved with middleware to avoid repeating code

If the user is unauth, throw an error, otherwise let the route execute

## Implementing protection middleware

### Protect function

```javascript
exports.protect = catchAsync(async (request, response, next) => {
    // Get the token and check if its there
    /** You can typically send a token as an http header, which can
  be accessed from `request.headers`
  The standard for supplying JWTs is below
  Authorization: Bearer [token here] */
    let token;
    if (
        request.headers.authorization &&
        request.headers.authorization.startsWith('Bearer')
    ) {
        token = request.headers.authorization.split(' ')[1];
    }
    if (!token) {
        return next(new AppError('Not logged in', 401));
    }
    console.log(token);

    // Validate and verify the token
    // We are going to promisify this method
    // If someone were to manipulate the token they would need to know the secret
    // to get it to pass verification
    const decoded = await promisify(jwt.verify)(token, process.env.JWT_SECRET);
    console.log(decoded);

    // If verified, check if user exists
    /**
     * If the user has changed his pw or the user has been deleted tokens
     * should be invalid
     */
    const freshUser = await User.findById(decoded.id);
    if (!freshUser) {
        return next(new AppError('User does not exist', 401));
    }

    // Check if user changed pw after token was issued
    if (freshUser.changedPasswordAfter(decoded.iat)) {
        return next(new AppError('Token is expired, please log in again', 401));
    }

    // The JWT and request has passed all checks, give access
    request.user = freshUser;
    next();
});
```

### In errorcontroller

```javascript
const handleJWTError = () =>
    new AppError('Invalid token. Please log in again', 401);

const handleExpiredTokenError = () =>
    new AppError('Invalid token. Please log in again', 401);
```

### In usermodel

```javascript
//
userSchema.methods.changedPasswordAfter = function (JWTTimeStamp) {
    if (this.passwordChangedAt) {
        const changedTimeStamp = parseInt(
            this.passwordChangedAt.getTime() / 1000,
            10
        );
        console.log(changedTimeStamp, JWTTimeStamp);
        return JWTTimeStamp < changedTimeStamp; // this will return true only if the password was changed after the token was issued
    }
    // false means not changed
    return false;
};
```

# User roles and permissions

Not all users should have rights to do certain things

So we should do authorization to confirm a user has the appropriate rights

```javascript
// if the roles dont match up, throw an error
// otherwise continue
// eslint-disable-next-line arrow-body-style
exports.restrictTo = (...roles) => {
    return (request, response, next) => {
        if (!roles.includes(request.user.role)) {
            return next(new AppError('Insufficient rights', 403));
        }

        next();
    };
};
```

# Resetting password

-   User sends POST request to a forgot password route
-   A token is sent to their email
-   User sends token email and new password to update password

Never store the reset token plainly in the database

```javascript
userSchema.methods.createPasswordResetToken = function () {
    const resetToken = crypto.randomBytes(32).toString('hex');

    this.passwordResetToken = crypto
        .createHash('sha256')
        .update(resetToken)
        .digest('hex');

    this.passwordResetExpires = Date.now() + 10 * 60 * 1000;

    return resetToken;
};
```

```javascript
exports.forgotPassword = catchAsync(async (request, response, next) => {
    // 1) Get user based on POSTed email
    const user = await User.findOne({ email: request.body.email });
    if (!user) {
        return next(new AppError('No user with that email address', 404));
    }
    // 2) Generate random reset token
    const resetToken = user.createPasswordResetToken();
    // we've updated the current document in our model but we still need to save it to db
    // you have to turn off validators for this particular instance
    await user.save({ validateBeforeSave: false });

    // 3) Send it back as an email
});
```
